Hue

cjstep opened the pull request:
This is a formula for Hue, a nice web interface for interacting with Hadoop stack applications.

The build takes a while - about 8 minutes on my 2014 Macbook Air.  The installed software is about 200 MB.  I'm going to look into bottling - this formula just has build-from-source for right now.

Happy to take any suggestions on formatting, installation method, etc.

Thanks!

- Chris

mistydemeo commented:
>  I'm going to look into bottling - this formula just has build-from-source for right now.

Don't worry about that - the build bots automatically build binaries.

mistydemeo commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,164 @@
+
+  def install
+    # patched the Makefile vars to install to ${PREFIX}/libexec
+    ohai "Building Hue from source - this process will take several minutes ..."
```
No need to include this.

mistydemeo commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,164 @@
+
+  test do
+
+    ohai "starting the Hue server in local mode"
```
No need for `ohai`s in the test block either.

mistydemeo commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,164 @@
+    system "make", "-j", "1", "install", "PREFIX=#{prefix}"
+
+    ohai "Setting up the default configuration file"
+    cp "desktop/conf.dist/hue.ini" "#{libexec}/desktop/conf/hue.ini"
```
This can be:

```ruby
(libexec/"desktop/conf").install "desktop/conf.dist/hue.ini"
```

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+  desc "Hue is a Web interface for analyzing data with Apache Hadoop."
+  homepage "http://gethue.com/"
+  url "https://github.com/cloudera/hue/archive/release-3.8.1.tar.gz"
+  # version "3.8.1" # redundant with version scanned from URL
```
Please remove this comment.

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+  # version "3.8.1" # redundant with version scanned from URL
+  sha256 "582777f567b9f4a34e4ce58cfd5ed24aff15f3e02f193e7990329d8cb0161a8a"
+
+  # system dependencies - either Homebrew or Mac OS versions
```
Please remove this comment.

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+
+  # system dependencies - either Homebrew or Mac OS versions
+  depends_on :java
+  depends_on :python
```
Adding a `:python` dependency introduces a hard dependency on Homebrew python for bottles. If it's not important which Python this is built against, please remove the `:python` dependency so it will always be built against system Python.

cjstep replied:
Done.

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+  depends_on :mysql
+
+  depends_on "maven" => :build
+  depends_on "openssl" => :optional
```
This needs to be mandatory or you need to find a way to tell Hue's build scripts not to build against openssl.

cjstep replied:
I made the dependency on "openssl" the default, and added an option to allow the use of system installed OpenSSL

```
  option "with-system-openssl", "Build with the Mac OS system OpenSSL instead of the (more secure) latest Homebrew version"
  depends_on "openssl" unless build.with? "system-openssl"
```

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+diff --git a/desktop/core/ext-py/tablib-develop/setup.py b/desktop/core/ext-py/tablib-develop/setup.py
+index fb16546..6f3bcfe 100755
+--- a/desktop/core/ext-py/tablib-develop/setup.py
++++ b/desktop/core/ext-py/tablib-develop/setup.py
```
Why is this patch necessary and has it been submitted upstream?

romainr replied:
Would Hue need this or a rebase on latest tablib? (that way you don't need this here)

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+diff --git a/desktop/core/ext-py/parquet-python/parquet/bitstring.py b/desktop/core/ext-py/parquet-python/parquet/bitstring.py
+index ab807a2..7203c94 100644
+--- a/desktop/core/ext-py/parquet-python/parquet/bitstring.py
++++ b/desktop/core/ext-py/parquet-python/parquet/bitstring.py
```
Same question here.

cjstep replied:
[Patch](https://github.com/cloudera/hue/pull/206) has been submitted upstream.

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+ # Installation directory
+ PREFIX ?= /usr/local
+-INSTALL_DIR ?= $(PREFIX)/hue
++INSTALL_DIR ?= $(PREFIX)/libexec
```
Looks like you could pass this on the command line instead of patching, i.e. `make install PREFIX=#{prefix} INSTALL_DIR=#{prefix}/libexec`.

cjstep replied:
Yeah, that was the first thing I tried.  When I set INSTALL_DIR from the command line, I get the following error when the first Hue web submodule is installed:

```
--- Making built distribution at /private/tmp/hue20150531-47630-1au188t/hue-release-3.8.1/desktop/core/build/bdist/desktop-3.8.1
--- Install built distribution for desktop at /usr/local/Cellar/hue/3.8.1/libexec
ERROR: /usr/local/Cellar/hue/3.8.1/libexec not empty. Cowardly refusing to continue.
make[2]: *** [install-bdist] Error 1
make[1]: *** [.recursive-install-bdist/core] Error 2
make: *** [install-desktop] Error 2
```

It looks like this is happening inside this target:

https://github.com/cloudera/hue/blob/master/Makefile.sdk#L245

So, I'd need to patch anyways.

I'm going to open an issue with Hue.  For now, patching the INSTALL_DIR variable instead of setting it on the command line works fine.

tdsmith commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,157 @@
+
+  def install
+    # patched the Makefile vars to install to ${PREFIX}/libexec
+    system "make", "-j", "1", "install", "PREFIX=#{prefix}"
```
Use `ENV.deparallelize` instead of passing `-j1` directly to make.

cjstep replied:
Done.

cjstep commented:
@tdsmith Thanks for all of the comments!  I'm working on a new revision, and will be back to reply soon.

DomT4 commented:
Any update here?

DomT4 commented:
Ping?

cjstep commented:
Hey Everyone,

Sorry for the delay!  I pushed a new commit tonight.

In addition to fixes for the issues above, I included:

* a [patch](https://github.com/cloudera/hue/issues/205) for a known issue with Hue's older version of pyOpenSSL
* a workaround for a change in the SASL include path location in Mavericks and higher

I'm still writing up the "tablib" pull request - I expect I'll have that out tomorrow or Thursday.

Cheers,

- Chris

DomT4 commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,172 @@
+  url "https://github.com/cloudera/hue/archive/release-3.8.1.tar.gz"
+  sha256 "582777f567b9f4a34e4ce58cfd5ed24aff15f3e02f193e7990329d8cb0161a8a"
+
+  option "with-system-openssl", "Build with the Mac OS system OpenSSL instead of the (more secure) latest Homebrew version"
```
We don't support this option in the core. It must use Homebrew's OpenSSL.

cjstep replied:
Removed option to install with system OpenSSL.

DomT4 commented:
> a patch for a known issue with Hue's older version of pyOpenSSL

Thanks for filing this upstream. Appreciate that. Hopefully we'll get a status update on that from them relatively soon - pyOpenSSL is fun to build nicely as intended/desired even when using the latest version, let alone when not :crying_cat_face:.

cjstep commented:
In the latest commit, I removed:

* the option to build with system OpenSSL.
* the "tablib" patch.  This has already been fixed in the dev branch of the tablib library.  As it turns out, the  issue only causes messages during the library's build, but doesn't cause the install to fail nor affect runtime.  

Also, Cloudera accepted the "parquet" and "pyOpenSSL" patches, so I'll be able to remove those from the formula on Hue's next release.

\- Chris

mikemcquaid commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,99 @@
+  depends_on "maven" => :build
+  depends_on "openssl"
+
+  patch :DATA
```
Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

mikemcquaid commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,99 @@
+
+  def install
+    ENV.append_to_cflags "-I" + `xcrun --show-sdk-path`.strip + "/usr/include/sasl" if MacOS.version >= :mavericks
+    ENV.deparallelize
```
Please file an upstream issue about this.

mikemcquaid replied:
Ping on this.

cjstep replied:
@mikemcquaid Just to make sure I understood which line you are referring to - are you asking if I can file an upstream issue on the build not working in parallel?

mikemcquaid replied:
Yes, please.

cjstep commented:
In this last commit, I removed the INSTALL_DIR patch that was fixing [this issue] (https://github.com/Homebrew/homebrew/pull/40150#discussion_r31401258).  The patch was just changing the install directory.  By default, Hue installs to "/usr/local/hue/...".  Now via Homebrew it installs to "libexec/hue/...", which is fine.

The remaining two patches were submitted upstream, and have been merged:

https://github.com/cloudera/hue/pull/206
https://github.com/cloudera/hue/pull/207

Added a comment to the formula with links and short notes.

\- Chris

mikemcquaid commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,94 @@
+  # (> 3.8.1) without them.
+  #
+  # Syntax error in 'parquet-python' library:
+  #   https://github.com/cloudera/hue/pull/206
```
Would it be possible to use `https://github.com/cloudera/hue/pull/206.patch` and `https://github.com/cloudera/hue/pull/207.patch` as `patch do` URLs instead?

cjstep replied:
I'll try it out.

cjstep replied:
Done.

mikemcquaid commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,94 @@
+  patch :DATA
+
+  def install
+    ENV.append_to_cflags "-I" + `xcrun --show-sdk-path`.strip + "/usr/include/sasl" if MacOS.version >= :mavericks
```
Interpolate the string here, thanks.

mikemcquaid replied:
Also you can get this from the `Xcode` module instead.

cjstep replied:
Good call!  Used the "#{MacOS.sdk_path}" variable instead.

cjstep commented:
Thanks everyone for the comments.  I'm still getting up to speed on Ruby + Homebrew best practices.  Anything else I can do to improve future formulas, please let me know.

dunn commented:
Looks like 3.9.0 is out. If you update this PR and push, it'll run another build.

bfontaine commented on Library/Formula/hue.rb:
```diff
@@ -0,0 +1,57 @@
+class Hue < Formula
+  desc "Hue is a Web interface for analyzing data with Apache Hadoop."
+  homepage "http://gethue.com/"
```
You can use `https://` here.

romainr replied:
Good point but gethue.com does not have https yet, it will break the page on modern browsers

bfontaine replied:
Right it breaks the page.

bfontaine commented:
Ping?

bfontaine commented:
Closing due to no response; feel free to open a new PR that passes the tests; thanks!
